### ISSUE TITLE
Examples about complex @BelongToMany associations in the doc

### ISSUE THREAD

**Issue by BenjaminDish** [RobinBuschmann/sequelize-typescript#412]
Hello,

This request is not an issue on the code but could be considered as a lack in the docs. I'm trying to use the @BelongsToMany tag do define a relationship between two entities, but I can't find how to do it really, using your documentation and the SequelizeJs documentation. My relation allows me to get a Map<string, CalculProgramme>. So I have an extra string attribute in the association (+ the 2 foreign keys)

I have an existing Postgresql database with the association already made.

Here is the schema :
```
CREATE TABLE programme
(
  id bigint NOT NULL,
  CONSTRAINT programme_pkey PRIMARY KEY (id)
);

CREATE TABLE calcul_programme (
  id bigint NOT NULL,
  CONSTRAINT calcul_programme_pkey PRIMARY KEY (id)
);

CREATE TABLE programme_calcul_programme (
  programme_id bigint NOT NULL,
  calculs_id bigint NOT NULL,
  calculs_key character varying(255) NOT NULL,
  CONSTRAINT programme_calcul_programme_pkey PRIMARY KEY (programme_id, calculs_id),
  CONSTRAINT fk_programme_calcul_programme_programme FOREIGN KEY (programme_id) REFERENCES programme (id),
  CONSTRAINT fk_programme_calcul_programme_calculs FOREIGN KEY (calculs_id) REFERENCES calcul_programme (id)
);
```

My classes are :

CalculProgramme.ts
```
@Table({tableName: 'calcul_programme', freezeTableName: true, underscored: true})
export class CalculProgramme extends Model<CalculProgramme> {
    @PrimaryKey   
    @AllowNull(false)
    @Column(DataType.BIGINT)
    id: number;

    /* Here is the hidden rest of the CalculProgramme's model */
}
```

Programme.ts
```
@Table({tableName: 'programme', freezeTableName: true, underscored: true})
export class Programme extends Model<Programme> {
    
    @PrimaryKey   
    @AllowNull(false)
    @Column(DataType.BIGINT)
    id: number;

    /* Here is the hidden rest of the Programme's model */

    /* HERE SHOULD BE THE ASSOCIATION I'M TRYING TO CONFIGURE*/
    @BelongsToMany(...)
    public calculs : {[key: string]: CalculProgramme};
}
```

**Comment by RobinBuschmann** [comment]
Hey @BenjaminDish , try it like so:

```ts
@Table(...)
export class Programme extends Model<Programme> {
  @BelongsToMany(() => CalculProgramme, () => ProgrammeCalculProgramme)
  calculPgrogrammes: CalculProgramme[];
}
```
```ts
@Table(...)
export class CalculProgramme extends Model<CalculProgramme> {
  @BelongsToMany(() => Programme, () => ProgrammeCalculProgramme)
  programmes: Programme[];
}
```

```ts
@Table(...)
export class ProgrammeCalculProgramme extends Model<ProgrammeCalculProgramme> {
  @PrimaryKey
  @ForeignKey(() => Programme)
  @Column
  programmeId: number;
  
  @PrimaryKey
  @ForeignKey(() =>  CalculProgramme)
  @Column
  calculProgrammeId: number;

  @Column
  calculsKey: string;
}
```

**Comment by BenjaminDish** [comment]
Hello @RobinBuschmann ,

Thank you for your answer, but the result I want is a bit diferent of what you propose. As you can see, I have a `calculs_key character varying(255) NOT NULL,` in my association. 

Indeed, the way I want the data to be retrieved in my model is :

```
@Table(...)
export class Programme extends Model<Programme> {
    @BelongsToMany(() => CalculProgramme, () => ProgrammeCalculProgramme)
    public calculs: { [key: string]: CalculProgramme };
}
```
where the key:string is the calculsKey of the association.

I am able to change the postgresql schema and maybe the @BelongsToMany association is not the appropriate one. Assuming that the only requirements is to have a model such like above, how would you define the model ?

**Comment by RobinBuschmann** [comment]
You could define a property getter additionally which resolves in such a map:
```ts
@Table(...)
export class Programme extends Model<Programme> {
  @BelongsToMany(() => CalculProgramme, () => ProgrammeCalculProgramme)
  calculProgrammes: CalculProgramme[];

  get calcus() {
    return (this.calculProgrammes || []).reduce((acc, calcul) => {
      // You can access the key via the through model which should be named like ProgrammeCalculProgramme
      acc[calcul.ProgrammeCalculProgramme.calculKey] = calcul;
      return acc;
    }, {})
  }
}
```
Does this help? :)

**Comment by BenjaminDish** [comment]
Hello Robin,

Thanks for your help.
I tried your piece of code but I can't test it properly. Indeed, the `acc[calcul.ProgrammeCalculProgramme.calculKey] = calcul;` row doesn't compile as ProgrammeCalculProgramme is not an attribute of CalculProgramme.

Almost I would like to have my 'calculs' field included in the json I return to my client. 

Sending data to the client should generate a JSON like that :

```
{  /* Instance of Programme */
    "id": 17,
     "calculs": {   
        "T1": {   /* The String key */
          /* Instance of CalculProgramme */
          "dateDerniereMiseAJour": "2018-07-18T16:19:57.289Z",
          "nbrLots": 0,
          "nbrLotsTvaReduite": 0,
          "surfaceMinimum": 0,
          "surfaceMoyenne": 0,
          "prixMinimum": 0,
          "prixMoyen": 0,
          "prixAuMetreCarreMoyen": 0,
          "surfaceMoyenneTvaReduite": 0,
          "prixMoyenTvaReduite": 0,
          "prixAuMetreCarreMoyenTvaReduite": 0,
          "id": 696773
        },
        "T2": {   /* The String key */
           /* Instance of CalculProgramme */
           "dateDerniereMiseAJour": "2018-07-18T16:19:57.289Z",
           "nbrLots": 43,
           "nbrLotsTvaReduite": 43,
           "surfaceMinimum": 25,
           "surfaceMoyenne": 43.03,
           "prixMinimum": 154123,
           "prixMoyen": 191094,
           "prixAuMetreCarreMoyen": 4460,
           "surfaceMoyenneTvaReduite": 43.03,
           "prixMoyenTvaReduite": 168138,
           "prixAuMetreCarreMoyenTvaReduite": 3926,
           "id": 696774
        }
     }
}
```

And when I receive such a JSON from my client, i would like to be able to save it (even the Calcul instances) using something like :

`programmeInstance.update(newData);`

Actually it seems that your workaround can't deal with it. Or maybe I didn't understand what you wanted to do.
And I feel like this would not be the 'sequelize way' to do it. Maybe Sequelize doesn't manage natively such associations, which is pretty sad, because often used in my code.

**Comment by BenjaminDish** [comment]
Hey Robin,

I finally gave up to apply such a structure to sequelize. I just added the key:string to the CalculProgramme model and used a Basic @BelongsToMany relationship between Programme and CalculProgramme.

Indeed, it seems that SequelizeJs is not done for this case.

Thank you for your help and for this project, he will soon be an important piece of my backend and I will follow, and maybe participate, to its evolutions.

Ciao
